### PR TITLE
csi: mount host's /etc/selinux in node plugins

### PR DESCRIPTION
**Pull request by humblec**
This commit introduces a new configuration option for
    ceph csi driver to enable hostpath mounting of /etc/selinux
    directory from the cluster node where csi plugin pods are
    running, which inturn help the csi driver to specify
    selinux-related mount options like context.
    
    Ref# https://github.com/ceph/ceph-csi/issues/2295
    
    The default value for this configuration is true and if cluster
    nodes are running without selinux enabled, an admin can deploy
    csi pods by specifying this option to `false` which skip the
    host path mounting for the csi pods.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by humblec** [comment]
@Madhu-1 ptal.. thanks

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml [hunk 1]
```diff
@@ -105,6 +105,9 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            - name: etc-selinux
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:108]: Can this be optional? In non-openshift or other non-selinux distros it wouldn't be needed

**Review by travisn — CHANGES_REQUESTED** [review]

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml (hunk 1 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:108]: @travisn that make sense,  anything specifically in your mind to make this optional mount?

Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:108]: Can you use a conditional like this? Then in the operator code that starts the CSI driver it would need to detect if in an environment where needed (or read it from a CSI setting).
```suggestion
            {{ if .SELinuxEnabled }}
            - name: etc-selinux
              mountPath: /etc/selinux
              readOnly: true
            {{ end }}
```

**Comment by travisn** [comment]
With #8418 note that the csi templates moved to a new folder under `pkg/operator/csi`

**Comment by leseb** [comment]
@humblec any plan to resume this?

**Comment by humblec** [comment]
> @humblec any plan to resume this?

@leseb @travisn  was planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept in this state.. sorry for missing to update the plan here.

**Comment by leseb** [comment]
> > @humblec any plan to resume this?
> 
> @leseb @travisn we were planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept this PR with this state.. sorry for missing to update the plan here.

No worries, what should we do with this PR? Thanks!

**Comment by humblec** [comment]
> > > @humblec any plan to resume this?
> > 
> > 
> > @leseb @travisn we were planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept this PR with this state.. sorry for missing to update the plan here.
> 
> No worries, what should we do with this PR? Thanks!

will update this PR addressing the comments and also an extra commit with the release template changes once the release is out  .. hope it works

**Comment by travisn** [comment]
> > > > @humblec any plan to resume this?
> > > 
> > > 
> > > @leseb @travisn we were planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept this PR with this state.. sorry for missing to update the plan here.
> > 
> > 
> > No worries, what should we do with this PR? Thanks!
> 
> will update this PR addressing the comments and also an extra commit with the release template changes once the release is out .. hope it works

Will these changes require the 3.4.1 driver? In other words, if someone is using an older version of the driver and this mount is still specified, will there be any issue? Can we test both version combinations with Rook and the 3.4.0/3.4.1 CSI driver?

**Comment by humblec** [comment]
> @humblec any plans to pursue this? Thanks

Let me share some details on what was happening on this meanwhile.. That said, this change is introduced here to follow up a change in ceph csi https://github.com/ceph/ceph-csi/issues/2295 which address an issue with long pod start delay if the pod has a big volume with many files in it. 
Considering this bind mount and this context population could help in OCP platform this was introduced as soon as the ceph csi PR merged.  On later experiments it has seen that, the selinux get enabled by default in OCP platform based on the container build. In that case, the fix we were targetting will work on OCP,  ie even without
bind mounting this share and contexts are populated as expected. When it comes to the upstream, this depend on where csi pods are getting spawned and if we see an selinux enabled system we have to bind mount, otherwise skip the mount. Indeed that was the suggestion here ( from @travisn ) and a good to have the fix. 
EOD, if systems are without selinx, the subjected fixed security context any way does not help , rather users will not be hitting pod startup delay due to 'selinux'

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml (hunk 1 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml]: yea, so in the latesr verison of this patch a condition is introduced which read from the csi setting.

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml [hunk 2]
```diff
@@ -105,6 +105,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            { { if .EnablePluginSelinuxHostMount } }
```
Comment (leseb) [inline pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:108]:
```suggestion
            {{ if .EnablePluginSelinuxHostMount }}
```

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml [hunk 3]
```diff
@@ -105,6 +105,11 @@ spec:
...
+            -  name: etc-selinux
+               mountPath: /etc/selinux
+               readOnly: true
+            { { end } }
```
Comment (leseb) [inline pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:112]:
```suggestion
            {{ end }}
```

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 4]
```diff
@@ -106,6 +106,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            { { if .EnablePluginSelinuxHostMount } }
```
Comment (leseb) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:109]:
```suggestion
            {{ if .EnablePluginSelinuxHostMount }}
```

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 5]
```diff
@@ -106,6 +106,11 @@ spec:
...
+            -  name: etc-selinux
+               mountPath: /etc/selinux
+               readOnly: true
+            { { end } }
```
Comment (leseb) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:113]:
```suggestion
            {{ end }}
```

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml (hunk 2 above)
Comment (humblec) [reply pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml]: Done.. thanks !

File: cluster/charts/rook-ceph/values.yaml [hunk 6]
```diff
@@ -66,6 +66,8 @@ csi:
   enableCephfsSnapshotter: true
   # set to false to disable deployment of snapshotter container in RBD provisioner pod.
   enableRBDSnapshotter: true
+  # set to false if the selinux is not enabled or unavailable in cluster nodes.
+  enablePluginSelinuxHostMount : true
```
Comment (Madhu-1) [inline cluster/charts/rook-ceph/values.yaml:70]: This need to be documented  in helm doc

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 7]
```diff
@@ -154,6 +154,10 @@ data:
 
   # (Optional) Allow starting unsupported ceph-csi image
   ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+
+  # (Optional) control the host mount of /etc/selinux for csi plugin pods.
```
Comment (Madhu-1) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:158]: don't we need to document when this needs to be enabled?

File: pkg/operator/ceph/csi/spec.go [hunk 8]
```diff
@@ -463,6 +468,7 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 		// get CephFS plugin tolerations and node affinity, defaults to common tolerations and node affinity if not specified
 		cephFSPluginTolerations := getToleration(r.opConfig.Parameters, cephFSPluginTolerationsEnv, pluginTolerations)
 		cephFSPluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, cephFSPluginNodeAffinityEnv, pluginNodeAffinity)
+
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:471]: unnecessary change not needed?

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml [hunk 9]
```diff
@@ -166,3 +171,6 @@ spec:
         - name: host-run-mount
           hostPath:
             path: /run/mount
+        - name: etc-selinux
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:174]: this should also have if check

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 10]
```diff
@@ -176,3 +181,6 @@ spec:
         - name: host-run-mount
           hostPath:
             path: /run/mount
+        - name: etc-selinux
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:184]: same here

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 11]
```diff
@@ -176,3 +181,6 @@ spec:
...
+          hostPath:
+            path: /etc/selinux
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:186]: will this path be present on all the distro and this path need to be configurable?

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (humblec) [reply cluster/charts/rook-ceph/values.yaml:70]: @travisn even in kubernetes, isnt it a good idea to default to `true` ?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 12]
```diff
@@ -154,6 +154,10 @@ data:
...
+  # CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
```
[the remark this answers is not in the document]
Comment (humblec) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:159]: yes, it is good to be enabled by default.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 11 above)
Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: I didnt get the question @Madhu-1 , can you please elaborate?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml (hunk 7 above)
Comment (humblec) [reply cluster/examples/kubernetes/ceph/operator-openshift.yaml:158]: we are defaulting to true here.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 11 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: Is the `/etc/selinux` path present on all the Linux distributions? if not deployment will fail?

Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: if selinux available system , it will be present. thus guaranteed to be working in platforms like openshift. For kube, we are defaulting to `false` and user can enable, if they are running selinux enabled cluster and if they want . does that answer @Madhu-1 ?

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (humblec) [reply cluster/charts/rook-ceph/values.yaml:70]: may be not, and good to be on false side here.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 13]
```diff
@@ -106,6 +106,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            {{ if .EnablePluginSelinuxHostMount }}
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:109]: @humblec https://github.com/ceph/ceph-csi/issues/2295 is reported only for cephfs. Do we need changes in rbd alss?

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 11 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: my question is is it always `/etc/selinux`? i see from some link its at `/etc/sysconfig/selinux` if it's not constant this need to be a configurable parameter for the user

Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: @Madhu-1 if you can share I can check the link, but to my knowledge its a symlink and we dont have to worry about it.

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (travisn) [reply cluster/charts/rook-ceph/values.yaml:70]: right, I'd still suggest false by default for non-openshift

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 13 above)
Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:109]: Enabling selinux relabelling is equally applicable for rbd too. so I have added it for rbd also.  @Madhu-1

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (humblec) [reply cluster/charts/rook-ceph/values.yaml:70]: agreed! patch carry the same. thanks.

**Comment by humblec** [comment]
The test failures looks spurious,

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 11 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: here are the links I found about it https://www.strongdm.com/docs/architecture/deployment/selinux https://www.omnisecu.com/gnu-linux/redhat-certified-engineer-rhce/security-enhanced-linux-selinux-configuration-file.php

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (travisn) [reply cluster/charts/rook-ceph/values.yaml:70]: Did you push the latest? It still looks true.

**Comment by humblec** [comment]
@travisn @leseb @Madhu-1 any further comments here ?

File: cluster/charts/rook-ceph/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (humblec) [reply cluster/charts/rook-ceph/values.yaml]: addressed, thanks.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 11 above)
Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:187]: as in the link https://www.omnisecu.com/gnu-linux/redhat-certified-engineer-rhce/security-enhanced-linux-selinux-configuration-file.php ```The SELinux configuration file is “/etc/sysconfig/selinux”, which is a symbolic link to “/etc/selinux/conf” file.``` , so should be fine.

File: pkg/operator/ceph/csi/spec.go [hunk 14]
```diff
@@ -316,6 +317,10 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 		tp.RBDPluginUpdateStrategy = rollingUpdate
 	}
 
+	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT", "true"), "false") {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:320]: To achieve a default of false, I believe we want this?
```suggestion
	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT", "false"), "true") {
```

File: pkg/operator/ceph/csi/spec.go [hunk 15]
```diff
@@ -316,6 +317,10 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
...
+		tp.EnableSelinuxHostMount = false
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go:321]: If we default to false unless the setting equals `true`, I think we need to switch this:
```suggestion
		tp.EnableSelinuxHostMount = true
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 14 above)
Comment (humblec) [reply pkg/operator/ceph/csi/spec.go]: correct .

File: pkg/operator/ceph/csi/spec.go (hunk 15 above)
Comment (humblec) [reply pkg/operator/ceph/csi/spec.go]: done.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml [hunk 16]
```diff
@@ -106,6 +106,11 @@ spec:
...
+            -  name: etc-selinux
```
Comment (travisn) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:110]: Not sure it's a problem, but there is an extra space on these lines
```suggestion
            - name: etc-selinux
```

**Review by travisn — CHANGES_REQUESTED** [review]
The changes look good, but the CI is failing, rbd pvs aren't being bound. I don't see a pod description of why the test pod is stuck pending, so it may be easier to test your changes locally than debugging the CI. We should also verify the changes on OpenShift with this setting enabled since we don't have that env in the CI.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml (hunk 16 above)
Comment (humblec) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml]: Thanks @travisn , thats been corrected. looks like tests are still failing.. so need to check on it..

**Comment by mykaul** [comment]
Is there an implicit assumption here that the selinux policy on the host and in the containers is the same? What if they are (for example), running different RHEL versions? Did we consult the selinux team on this?

**Comment by humblec** [comment]
> Is there an implicit assumption here that the selinux policy on the host and in the containers is the same? What if they are (for example), running different RHEL versions? 

Good question @mykaul , actually that was  the scenario where we have started with. That said, while we have  different `libmount` ( util-lnux)  version to be precise, thus similar to a situation of  different rhel version in container and host scenario we have seen that, the selinux contexts were not getting respected, thus not able to make the context population work properly.  So the host mount fix was introduced.  

Apart from above, have gone through the selinux code and from the selinux code reading what have been identified is that, this mount is actually inspected just for selinux enablement part. (via the `is_selinux_enabled` and `selinux_mnt` ( set by `init_selinuxmnt`->`verify_selinuxmnt `which make the decision based on `SELINUXMNT and OLDSELINUXMNT` variables [1]), thats where I was thinking this is safe to bind mount and also being its `readonly` mount , it should be fine. The other decision point which selinux make use of in the code  is based on `/sys`  which any way our node plugin pods are mounting from the host almost since start of csi driver. To summarize,  even if host and containers builds are different, it looks like we are good. 

>Did we consult the selinux team on this?

havent consulted selinux team specifically on this and only bit of code reading in specific areas and few reports from upstream users that, things are working with this bind mount and no complaints so far we have heard. However I can also check with selinux team and double confirm about this approach. 

[1] https://github.com/SELinuxProject/selinux/blob/d155b410d4bbc90d28f361b966f0429598da8188/libselinux/src/enabled.c

**Comment by leseb** [comment]
@humblec isn't `/sys/fs/selinux` sufficient for `is_selinux_enabled` to return true? Why is `/etc/selinux` adding to this detection? Also, I left a comment for clarification here https://github.com/ceph/ceph-csi/issues/2295#issuecomment-980048956, thanks!

**Comment by humblec** [comment]
> @humblec isn't `/sys/fs/selinux` sufficient for `is_selinux_enabled` to return true? Why is `/etc/selinux` adding to this detection? Also, I left a comment for clarification here [ceph/ceph-csi#2295 (comment)](https://github.com/ceph/ceph-csi/issues/2295#issuecomment-980048956), thanks!

selinux also check for the config in its internal code path, so this helps.

**Comment by humblec** [comment]
> The changes look good, but the CI is failing, rbd pvs aren't being bound. I don't see a pod description of why the test pod is stuck pending, so it may be easier to test your changes locally than debugging the CI. We should also verify the changes on OpenShift with this setting enabled since we don't have that env in the CI.

@travisn  So all tests are passing and also verified this in a kubernetes setup with the default value and verified PVCs are bound and working properly.  The OCP validation with flipping the value to `true` has not been done yet though.

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 17]
```diff
@@ -154,6 +154,10 @@ data:
 
   # (Optional) Allow starting unsupported ceph-csi image
   ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+
+  # (Optional) control the host mount of /etc/selinux for csi plugin pods.
+  # CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "true"
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml:159]: This is the same default, but perhaps easier to see the default than when it is commented out. 
```suggestion
  CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
```

File: cluster/examples/kubernetes/ceph/operator.yaml [hunk 18]
```diff
@@ -72,6 +72,10 @@ data:
 
   # (Optional) Allow starting unsupported ceph-csi image
   ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+
+  # (Optional) control the host mount of /etc/selinux for csi plugin pods.
+  # CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator.yaml:77]:
```suggestion
  CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
```

**Review by travisn — CHANGES_REQUESTED** [review]
To resolve the merge conflicts, note the following changes from #8740:
- `cluster/examples/kubernetes/ceph/` moved to `deploy/examples`
- `cluster/charts/` moved to `deploy/charts`

**Comment by humblec** [comment]
> To resolve the merge conflicts, note the following changes from #8740:
> 
>     * `cluster/examples/kubernetes/ceph/` moved to `deploy/examples`
> 
>     * `cluster/charts/` moved to `deploy/charts`

Rebased to master.. Thanks , ptal..

**Review by leseb — APPROVED** [review]

**Review by travisn — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]
